client/src/game/ui/selection/shapecontext.vue

```diff
@@ -19,85 +19,133 @@
 import { EventBus } from "@/game/event-bus";
 import { layerManager } from "@/game/layers/manager";
-import { Shape } from "@/game/shapes/shape";
 import { gameStore } from "@/game/store";
-import { inInitiative, initiativeStore } from "../initiative/store";
+import { cutShapes, deleteShapes, pasteShapes } from "../../shapes/utils";
+import { initiativeStore, inInitiative } from "../initiative/store";
+import { Vector } from "../../geom";
 
 @Component({
     components: {
         ContextMenu,
     },
+    computed: {
+        ...mapState("game", ["locations", "assets", "notes"]),
+    },
 })
 export default class ShapeContext extends Vue {
     visible = false;
     x = 0;
     y = 0;
-    shape: Shape | null = null;
     get activeLayer(): string {
         const layer = layerManager.getLayer();
         return layer === undefined ? "" : layer.name;
     }
-    open(event: MouseEvent, shape: Shape) {
+    open(event: MouseEvent) {
         this.visible = true;
         this.x = event.pageX;
         this.y = event.pageY;
-        this.shape = shape;
         this.$nextTick(() => (<HTMLElement>this.$children[0].$el).focus());
     }
     close() {
         this.visible = false;
-        this.shape = null;
     }
     getLayers() {
         return layerManager.layers.filter(l => l.selectable && (gameStore.IS_DM || l.playerEditable));
     }
     getActiveLayer() {
         return layerManager.getLayer();
     }
+    getCurrentLocation() {
+        return gameStore.locationName;
+    }
     getInitiativeWord() {
-        if (this.shape === null) return "";
-        return inInitiative(this.shape.uuid) ? "Show" : "Add";
+        const layer = layerManager.getLayer()!;
+        if (layer.selection.length === 1) {
+            return inInitiative(layer.selection[0].uuid) ? "Show" : "Add";
+        } else {
+            return layer.selection.every(shape => inInitiative(shape.uuid)) ? "Show" : "Add all to";
+        }
+    }
+    hasSingleShape(): boolean {
+        const layer = layerManager.getLayer()!;
+        return layer.selection.length === 1;
     }
     setLayer(newLayer: string) {
-        if (this.shape === null) return;
-        this.shape.moveLayer(newLayer, true);
+        const layer = layerManager.getLayer()!;
+        layer.selection.forEach(shape => shape.moveLayer(newLayer, true));
+        this.close();
+    }
+    setLocation(newLocation: string) {
+        const layer = layerManager.getLayer()!;
+        cutShapes();
+        // Request change to other location
+        socket.emit("Location.Change", newLocation);
+        socket.once("Location.Set", (_data: Partial<ServerLocation>) => {
+            socket.once("Client.Options.Set", (_options: ServerClient) => {
+                // TODO: Shapes are pasted to their original position instead of the position the camera is now centered
+                layer.selection = pasteShapes(new Vector(0, 0), layer.name);
```

Comment:
```suggestion
                layer.selection = pasteShapes(undefined, layer.name);
```
I understand the reason why you chose new Vector(0, 0) here, but it's an unnecessary allocation, that if left as undefined, will jump straight over the if (offset) line with the same end-effect.